### PR TITLE
Allowing auto-restarts on fail for Perfect and Sudden Death mod

### LINKED ISSUE #3380: Add an auto-restart option after failing for Perfect
In osu!stable, whenever you fail on Perfect, the game automatically retries. In osu!lazer, that doesn't happen. This could be an additional option that could be picked in the settings. It could work not only for Perfect but also SuddenDeath. 

~~I might be going a bit too far with those stable to lazer ideas~~

### PR DESCRIPTION
**Pull request by plankp**
Partially fixes #3380 since this does perform the auto-restart, but it is not optional.

### CODE REVIEW

**Comment by plankp** [comment]
Now fully addresses #3380: By default auto-restart for SD and PF is disabled, but can be enabled in settings under the mods section.

**Comment by smoogipoo** [comment]
I think this would work better if it was integrated into the game rather than being an option, tbh. Something like the restart button having a 3sec timer counting down which is cancelled upon pressing escape or the exit button.

File: osu.Game/Rulesets/Mods/IApplicableRestartOnFail.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// Represents a mod which can request to restart on fail.
+    /// </summary>
+    public interface IApplicableRestartOnFail : IApplicableMod
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/IApplicableRestartOnFail.cs]: How about combining this into `IApplicableFailOverride`? Make a property like `RestartOnFail` in it.

**Review by smoogipoo — DISMISSED** [review]

**Comment by peppy** [comment]
The fail doesn't actually happen with the current code. It doesn't stop the audio. The restart can be called multiple times before it actually completes. Was this actually tested?

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -289,6 +289,12 @@ private bool onFail()
             if (Beatmap.Value.Mods.Value.OfType<IApplicableFailOverride>().Any(m => !m.AllowFail))
                 return false;
 
+            if (Beatmap.Value.Mods.Value.OfType<IApplicableFailOverride>().Any(m => m.RestartOnFail))
+            {
+                Restart();
+                return false;
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: This should `return true` otherwise `onFail()` is going to be called multiple times.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
`osu-stable` does not auto restart for sudden death. was this an oversight or intentional change?

**Comment by plankp** [comment]
Intentional (extract from linked issue: *It could work not only for Perfect but also SuddenDeath.*), but I could change it back.

**Comment by peppy** [comment]
As long as it's a conscious change I guess this is fine for now. The idea is it will be a mod-specific setting in the future (when we have support for these).

**Comment by peppy** [comment]
See #6159.

**Pull request closed without merging**